src/stream/src/executor/backfill.rs

```diff
@@ -188,102 +264,135 @@ where
-                                    cur_barrier_upstream_processed_rows +=
-                                        chunk.cardinality() as u64;
-                                    if let Some(current_pos) = &current_pos {
+                                    self.metrics
+                                        .backfill_snapshot_read_row_count
+                                        .with_label_values(&[
+                                            upstream_table_id.to_string().as_str(),
+                                            self.actor_id.to_string().as_str(),
+                                        ])
+                                        .inc_by(cur_barrier_snapshot_processed_rows);
+
+                                    self.metrics
+                                        .backfill_upstream_output_row_count
+                                        .with_label_values(&[
+                                            upstream_table_id.to_string().as_str(),
+                                            self.actor_id.to_string().as_str(),
+                                        ])
+                                        .inc_by(cur_barrier_upstream_processed_rows);
+
+                                    // Update snapshot read epoch.
+                                    snapshot_read_epoch = barrier.epoch.prev;
+
+                                    self.progress.update(
+                                        barrier.epoch.curr,
+                                        snapshot_read_epoch,
+                                        total_snapshot_processed_rows,
+                                    );
+
+                                    // Persist state on barrier
+                                    Self::persist_state(
+                                        barrier.epoch,
+                                        &mut self.state_table,
+                                        false,
+                                        &current_pos,
+                                        &mut old_state,
+                                        &mut current_state,
+                                    )
+                                    .await?;
+
+                                    yield Message::Barrier(barrier);
+                                    // Break the for loop and start a new snapshot read stream.
+                                    break;
+                                }
+                                Message::Chunk(chunk) => {
+                                    // Buffer the upstream chunk.
+                                    upstream_chunk_buffer.push(chunk.compact());
+                                }
+                                Message::Watermark(watermark) => {
+                                    self.state_table
+                                        .update_watermark(watermark.val.clone(), false);
+                                    // Ignore watermark during backfill.
```

Comment:
Why didn't we forward `Message::Watermark` previously? Should I forward it?

Reply:
We shouldn't forward it, because backfill executors could generate some rows violating the watermark constraint.